sympy/core/basic.py

```diff
@@ -1284,8 +1311,27 @@ def has_free(self, *patterns):
         True
         >>> (x + y + 1).has_free(y + 1)
         True
-
         """
+        if not patterns:
+            return False
+        p0 = patterns[0]
+        if len(patterns) == 1 and iterable(p0) and not isinstance(p0, Basic):
+            # Basic can contain iterables (though not non-Basic, ideally)
+            # but don't encourage mixed passing patterns
+            raise ValueError(filldedent('''
+                Expecting 1 or more Basic args, not a single
+                non-Basic iterable. Don't forget to unpack
+                iterables: `eq.has_free(*patterns)`'''))
+        # try quick test first
+        try:
+            s = set(patterns)
+        except TypeError:
+            pass  # patterns had a non-hashable element
```

Comment:
As I've said before I would prefer not to catch `TypeError`. There are too many places that do this. A `TypeError` should almost always mean a bug and so the error should not generally be caught.

Reply:
Please make a suggestion since there is no other error to catch and the scope has been limited to a single obvious case of why it might be raised. 

The reason this is here is because if we can make a set of the items and this test passes then it is a quick exit. And if not (either unhashable items or pattern not found) then we go to the next step where unhashable things are thrown out. In case of unhashable I don't want to let this raise for the user because the next step will ignore them (or could its own TypeError).

Reply:
Why we can't we just let this TypeError bubble up?

Reply:
I guess we can see what fails...will make the change